gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1426 @@
+      outputStream.close();
+    } catch (IOException e) {
+      log.log(Level.WARNING, "IOException encountered when sending response.", e);
+    }
+  }
+
+  /**
+   * Decodes content of the HttpRequest.
+   */
+  private static String decodeContent(Headers headers, InputStream inputStream) throws IOException {
+    List<String> contentEncoding = headers.get("Content-encoding");
+    InputStream input = inputStream;
+    try {
+      if (contentEncoding != null && !contentEncoding.isEmpty()) {
+        String encoding = contentEncoding.get(0);
+        if (SUPPORTED_COMPRESSION_ENCODINGS.contains(encoding)) {
+          input = new GZIPInputStream(inputStream);
+        } else if (!encoding.equals("identity")) {
+          throw new IOException(
+              "The request has the following unsupported HTTP content encoding: " + encoding);
+        }
+      }
+      return new String(ByteStreams.toByteArray(input), StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      throw new IOException("Exception encountered when decoding request content.", e);
+    }
+  }
+
+  /**
+   * Generates a JSON response.
+   *
+   * @param context managedZones | projects | rrsets | changes
+   */
+  static Response toListResponse(List<String> serializedObjects, String context, String pageToken,
+      boolean includePageToken) {
+    // start building response
+    StringBuilder responseBody = new StringBuilder();
+    responseBody.append("{\"").append(context).append("\": [");
+    Joiner.on(",").appendTo(responseBody, serializedObjects);
+    responseBody.append("]");
+    // add page token only if exists and is asked for
+    if (pageToken != null && includePageToken) {
+      responseBody.append(",\"nextPageToken\": \"").append(pageToken).append("\"");
+    }
+    responseBody.append("}");
+    return new Response(HTTP_OK, responseBody.toString());
+  }
+
+  /**
+   * Prepares DNS records that are created by default for each zone.
+   */
+  private static ImmutableList<RrsetWrapper> defaultRecords(ManagedZone zone) {
+    ResourceRecordSet soa = new ResourceRecordSet();
+    soa.setTtl(21600);
+    soa.setName(zone.getDnsName());
+    soa.setRrdatas(ImmutableList.of(
+        // taken from the service
+        "ns-cloud-c1.googledomains.com. cloud-dns-hostmaster.google.com. 0 21600 3600 1209600 312"
+    ));
+    soa.setType("SOA");
+    ResourceRecordSet ns = new ResourceRecordSet();
+    ns.setTtl(21600);
+    ns.setName(zone.getDnsName());
+    ns.setRrdatas(zone.getNameServers());
+    ns.setType("NS");
+    RrsetWrapper nsWrapper = new RrsetWrapper(ns);
+    RrsetWrapper soaWrapper = new RrsetWrapper(soa);
+    ImmutableList<RrsetWrapper> results = ImmutableList.of(nsWrapper, soaWrapper);
+    nsWrapper.setId(getUniqueId(results));
+    soaWrapper.setId(getUniqueId(results));
+    return results;
+  }
+
+  /**
+   * Returns a list of four nameservers randomly chosen from the predefined set.
+   */
+  static List<String> randomNameservers() {
```

Comment:
Out of curiosity, any reason for returning 4 nameservers?  Do you think users will benefit more from the randomness than than having a constant set of name servers (which might make their tests more reproducible).